Bottom: 064fb858fa40066728b05dc76c85196499aec143
Top:    8689477a27f06497fae45286c91b4161ef191ae3
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-02 05:35:36 +0000

Refresh of improve-gl-stuff

---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index bb21ba9..eb1044f 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -69,12 +69,7 @@ hidgl_init_triangle_array (triangle_buffer *buffer)
   glEnableClientState (GL_VERTEX_ARRAY);
   glGenBuffers (1, &buffer->vbo_name);
   glBindBuffer (GL_ARRAY_BUFFER, buffer->vbo_name);
-                                                 /* HUGE BUFFER */
-//  glBufferData (GL_ARRAY_BUFFER, sizeof (GLFloat) * 8192 * 2 * 3, NULL, GL_STATIC_DRAW);
   glBufferData (GL_ARRAY_BUFFER, TRIANGLE_ARRAY_BYTES, NULL, GL_STATIC_DRAW);
-//  glBufferSubData (GL_ARRAY_BUFFER, offsetInByte, SizeInBytes, &pvertices);
-//  glBindBuffer (GL_ARRAY_BUFFER, &buffer->vbo_name);
-//  glDeleteBuffers (1, &buffer->vbo_name);
 
   buffer->triangle_array = glMapBuffer (GL_ARRAY_BUFFER, GL_WRITE_ONLY);
 
diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 7738098..17ecae2 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -1077,6 +1077,11 @@ ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
   hidgl_flush_triangles (&buffer);
   glPopMatrix ();
 
+
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
+
   draw_grid ();
 
   hidgl_init_triangle_array (&buffer);
@@ -1097,6 +1102,10 @@ ghid_port_drawing_area_expose_event_cb (GtkWidget * widget,
 
   hidgl_flush_triangles (&buffer);
 
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
+
   if (gdk_gl_drawable_is_double_buffered (pGlDrawable))
     gdk_gl_drawable_swap_buffers (pGlDrawable);
   else
diff --git a/src/hid/gtk/gui-pinout-preview.c b/src/hid/gtk/gui-pinout-preview.c
index 137471f..bebaf96 100644
--- a/src/hid/gtk/gui-pinout-preview.c
+++ b/src/hid/gtk/gui-pinout-preview.c
@@ -236,6 +236,10 @@ ghid_pinout_preview_expose (GtkWidget * widget, GdkEventExpose * ev)
   hidgl_flush_triangles (&buffer);
   glPopMatrix ();
 
+  glUnmapBuffer (GL_ARRAY_BUFFER);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+  glDeleteBuffers (1, &buffer.vbo_name);
+
   if (gdk_gl_drawable_is_double_buffered (pGlDrawable))
     gdk_gl_drawable_swap_buffers (pGlDrawable);
   else
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index c8e9d85..c51a0c9 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -35,9 +35,15 @@
 
 #include <gtk/gtk.h>
 
+/* The Linux OpenGL ABI 1.0 spec requires that we define
+ * GL_GLEXT_PROTOTYPES before including gl.h or glx.h for extensions
+ * in order to get prototypes:
+ *   http://www.opengl.org/registry/ABI/
+ */
 #ifdef ENABLE_GL
-#  include <gtk/gtkgl.h>
+#  define GL_GLEXT_PROTOTYPES 1
 #  include <GL/gl.h>
+#  include <gtk/gtkgl.h>
 #  include "hid/common/hidgl.h"
 #endif
